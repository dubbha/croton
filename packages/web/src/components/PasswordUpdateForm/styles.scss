.btn {
  width: 100%;
}

form {
  width: 70%;
  margin-bottom: 90px;
}

