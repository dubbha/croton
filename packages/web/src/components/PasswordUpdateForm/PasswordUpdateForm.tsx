import React, { useState, FormEvent, useEffect } from 'react';
import {
  Form,
  SubmitButton,
  ErrorAlert,
  InfoAlert,
  AlertPlaceholder,
  LoadingSpinner,
} from 'elements';
import './styles.scss';

type Props = {
  isLoading: boolean;
  error: string | null;
  info: string | null;
  onSubmit: (email: string) => void;
};

export const PasswordUpdateForm = ({
  isLoading,
  error,
  info,
  onSubmit,
}: Props) => {
  const [isValid, setIsValid] = useState(false);
  const [password, setPassword] = useState('');
  const [passwordMatch, setPasswordMatch] = useState('');
  const [isPasswordMatch, setIsPasswordMatch] = useState(true);

  useEffect(() => {
    setIsPasswordMatch(true);
  }, [password, passwordMatch]);

  useEffect(() => {
    setIsValid(!!password && !!passwordMatch);
  }, [password, passwordMatch, isPasswordMatch]);

  const handleSubmit = (e: FormEvent) => {
    e.preventDefault();
    if (password !== passwordMatch) {
      setIsPasswordMatch(false);
      return;
    }
    onSubmit(password);
  };

  return (
    <Form onSubmit={handleSubmit}>
      {error && <ErrorAlert>{error}</ErrorAlert>}
      {info && <InfoAlert>{info}</InfoAlert>}
      {!error && !info && <AlertPlaceholder />}
      <Form.Group>
        <Form.Label>New Password</Form.Label>
        <Form.Control
          type="password"
          placeholder="Enter new password"
          value={password}
          onChange={e => setPassword(e.target.value)}
          data-testid="passwordUpdateForm__password"
        />
      </Form.Group>

      <Form.Group>
        <Form.Label>Repeat Password</Form.Label>
        <Form.Control
          type="password"
          placeholder="Repeat password"
          value={passwordMatch}
          onChange={e => setPasswordMatch(e.target.value)}
          data-testid="passwordUpdateForm__passwordMatch"
        />
        {!isPasswordMatch && (
          <Form.Text id="passwordHelpBlock" className="password-not-match">
            Your passwords must match.
          </Form.Text>
        )}
      </Form.Group>

      <SubmitButton disabled={!isValid || isLoading}>
        <div className="spinner-container">
          {isLoading && <LoadingSpinner />}
        </div>
        <span>Update Password</span>
      </SubmitButton>
    </Form>
  );
};
