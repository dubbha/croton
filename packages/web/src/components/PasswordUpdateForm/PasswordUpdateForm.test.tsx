import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import { PasswordUpdateForm } from './PasswordUpdateForm';

describe('components/PasswordUpdateForm', () => {
  const fn = jest.fn();

  const props = {
    isLoading: false,
    error: null,
    info: null,
    onSubmit: fn
  };

  afterEach(() => {
    fn.mockReset();
  });

  it('should render successfully', () => {
    const { container } = render(<PasswordUpdateForm {...props} />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render submit button disabled by default', () => {
    const { container, getByTestId } = render(
      <PasswordUpdateForm {...props} />
    );
    expect(container.firstChild).toMatchSnapshot();

    const submitButton = getByTestId('submitButton');
    expect(submitButton).toBeDisabled();
  });

  it('should call onSubmit prop on submit', () => {
    const { container, getByTestId } = render(
      <PasswordUpdateForm {...props} />
    );
    expect(container.firstChild).toMatchSnapshot();

    const passwordInput = getByTestId('passwordUpdateForm__password');
    const passwordMatchInput = getByTestId('passwordUpdateForm__passwordMatch');
    const submitButton = getByTestId('submitButton');

    fireEvent.change(passwordInput, { target: { value: 'pass' } });
    fireEvent.change(passwordMatchInput, { target: { value: 'pass' } });

    expect(submitButton).toBeEnabled();

    fireEvent.click(submitButton);
    expect(fn).toBeCalled();
  });

  it('should enforce matching passwords', () => {
    const { getByTestId, getByText } = render(
      <PasswordUpdateForm {...props} />
    );

    const passwordInput = getByTestId('passwordUpdateForm__password');
    const passwordMatchInput = getByTestId('passwordUpdateForm__passwordMatch');
    const submitButton = getByTestId('submitButton');

    fireEvent.change(passwordInput, { target: { value: 'pass' } });
    fireEvent.change(passwordMatchInput, { target: { value: 'PASS' } });

    expect(submitButton).toBeEnabled();

    fireEvent.click(submitButton);

    expect(fn).not.toBeCalled();
    expect(getByText('Your passwords must match.')).toBeTruthy();
  });

  it('should render spinner when loading', () => {
    const { container } = render(<PasswordUpdateForm {...props} isLoading />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should display error alert', () => {
    const { getByTestId } = render(
      <PasswordUpdateForm {...props} error="ERROR" />
    );
    const errorAlert = getByTestId('errorAlert');
    expect(errorAlert.textContent).toBe('ERROR');
  });

  it('should display info alert', () => {
    const { getByTestId } = render(
      <PasswordUpdateForm {...props} info="INFO" />
    );
    const infoAlert = getByTestId('infoAlert');
    expect(infoAlert.textContent).toBe('INFO');
  });
});
