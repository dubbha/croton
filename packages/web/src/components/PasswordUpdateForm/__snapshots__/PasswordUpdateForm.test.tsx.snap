// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/PasswordUpdateForm should call onSubmit prop on submit 1`] = `
<form
  class=""
>
  <div
    style="height: 50px; margin-bottom: 16px;"
  />
  <div
    class="form-group"
  >
    <label
      class="form-label"
    >
      New Password
    </label>
    <input
      class="form-control"
      data-testid="passwordUpdateForm__password"
      placeholder="Enter new password"
      type="password"
      value=""
    />
  </div>
  <div
    class="form-group"
  >
    <label
      class="form-label"
    >
      Repeat Password
    </label>
    <input
      class="form-control"
      data-testid="passwordUpdateForm__passwordMatch"
      placeholder="Repeat password"
      type="password"
      value=""
    />
  </div>
  <button
    class="btn btn-primary"
    data-testid="submitButton"
    disabled=""
    type="submit"
  >
    <div
      class="spinner-container"
    />
    <span>
      Update Password
    </span>
  </button>
</form>
`;

exports[`components/PasswordUpdateForm should render spinner when loading 1`] = `
<form
  class=""
>
  <div
    style="height: 50px; margin-bottom: 16px;"
  />
  <div
    class="form-group"
  >
    <label
      class="form-label"
    >
      New Password
    </label>
    <input
      class="form-control"
      data-testid="passwordUpdateForm__password"
      placeholder="Enter new password"
      type="password"
      value=""
    />
  </div>
  <div
    class="form-group"
  >
    <label
      class="form-label"
    >
      Repeat Password
    </label>
    <input
      class="form-control"
      data-testid="passwordUpdateForm__passwordMatch"
      placeholder="Repeat password"
      type="password"
      value=""
    />
  </div>
  <button
    class="btn btn-primary"
    data-testid="submitButton"
    disabled=""
    type="submit"
  >
    <div
      class="spinner-container"
    >
      <span
        aria-hidden="true"
        class="spinner-border spinner-border-sm"
        role="status"
      />
    </div>
    <span>
      Update Password
    </span>
  </button>
</form>
`;

exports[`components/PasswordUpdateForm should render submit button disabled by default 1`] = `
<form
  class=""
>
  <div
    style="height: 50px; margin-bottom: 16px;"
  />
  <div
    class="form-group"
  >
    <label
      class="form-label"
    >
      New Password
    </label>
    <input
      class="form-control"
      data-testid="passwordUpdateForm__password"
      placeholder="Enter new password"
      type="password"
      value=""
    />
  </div>
  <div
    class="form-group"
  >
    <label
      class="form-label"
    >
      Repeat Password
    </label>
    <input
      class="form-control"
      data-testid="passwordUpdateForm__passwordMatch"
      placeholder="Repeat password"
      type="password"
      value=""
    />
  </div>
  <button
    class="btn btn-primary"
    data-testid="submitButton"
    disabled=""
    type="submit"
  >
    <div
      class="spinner-container"
    />
    <span>
      Update Password
    </span>
  </button>
</form>
`;

exports[`components/PasswordUpdateForm should render successfully 1`] = `
<form
  class=""
>
  <div
    style="height: 50px; margin-bottom: 16px;"
  />
  <div
    class="form-group"
  >
    <label
      class="form-label"
    >
      New Password
    </label>
    <input
      class="form-control"
      data-testid="passwordUpdateForm__password"
      placeholder="Enter new password"
      type="password"
      value=""
    />
  </div>
  <div
    class="form-group"
  >
    <label
      class="form-label"
    >
      Repeat Password
    </label>
    <input
      class="form-control"
      data-testid="passwordUpdateForm__passwordMatch"
      placeholder="Repeat password"
      type="password"
      value=""
    />
  </div>
  <button
    class="btn btn-primary"
    data-testid="submitButton"
    disabled=""
    type="submit"
  >
    <div
      class="spinner-container"
    />
    <span>
      Update Password
    </span>
  </button>
</form>
`;
